lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1308,30 +1373,24 @@ void Differentiation::run() {
     context.addInvoker(&witness);
   }
 
-  // Register all the `differentiable_function` instructions in the module that
-  // trigger differentiation.
+  // Register all the `differentiable_function` and `linear_function`
+  // instructions in the module that trigger differentiation.
   for (SILFunction &f : module) {
     for (SILBasicBlock &bb : f) {
       for (SILInstruction &i : bb) {
-        if (auto *dfi = dyn_cast<DifferentiableFunctionInst>(&i))
-          context.addDifferentiableFunctionInstToWorklist(dfi);
-        // Reject uncanonical `linear_function` instructions.
-        // FIXME(SR-11850): Add support for linear map transposition.
-        else if (auto *lfi = dyn_cast<LinearFunctionInst>(&i)) {
-          if (!lfi->hasTransposeFunction()) {
-            astCtx.Diags.diagnose(
-                lfi->getLoc().getSourceLoc(),
-                diag::autodiff_conversion_to_linear_function_not_supported);
-            errorOccurred = true;
-          }
+        if (auto *dfi = dyn_cast<DifferentiableFunctionInst>(&i)) {
+          context.getDifferentiableFunctionInstWorklist().push_back(dfi);
+        } else if (auto *lfi = dyn_cast<LinearFunctionInst>(&i)) {
+          context.getLinearFunctionInstWorklist().push_back(lfi);
         }
```

Comment:
`test/AutoDiff/SILOptimizer/differentiation_diagnostics.swift` fails because the ``conversion to '@differentiable(linear)' function type is not yet supported)`` diagnostic was removed:
https://ci.swift.org/job/swift-PR-Linux/22592/console
```
15:00:28 ******************** TEST 'Swift(linux-x86_64) :: AutoDiff/SILOptimizer/differentiation_diagnostics.swift' FAILED ********************
15:00:28 Script:
15:00:28 --
15:00:28 : 'RUN: at line 1';   /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/swift-linux-x86_64/bin/swift-frontend -target x86_64-unknown-linux-gnu  -module-cache-path '/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache' -swift-version 4  -ignore-module-source-info  -typo-correction-limit 10  -emit-sil -verify /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift/test/AutoDiff/SILOptimizer/differentiation_diagnostics.swift
15:00:28 --
15:00:28 Exit Code: 1
15:00:28 
15:00:28 Command Output (stderr):
15:00:28 --
15:00:28 
/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift/test/AutoDiff/SILOptimizer/differentiation_diagnostics.swift:879:4: error: expected error not produced
15:00:28 // expected-error @+1 {{conversion to '@differentiable(linear)' function type is not yet supported}}
15:00:28 ~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Since this PR is only a first step towards complete linear map transposition, I think it's best to add a `-Xllvm -enable-experimental-linear-map-transposition` feature flag. If the flag is not enabled, we should produce the old diagnostic.

----

You can declare this flag near the top of the file:

```cpp
/// This flag enables experimental `@differentiable(linear)` function
/// transposition.
static llvm::cl::opt<bool> EnableExperimentalLinearMapTransposition(
    "enable-experimental-linear-map-transposition", llvm::cl::init(false));
```

Then check the flag when visiting `linear_function` instructions, to decide whether to diagnose:
```suggestion
        } else if (auto *lfi = dyn_cast<LinearFunctionInst>(&i)) {
          // If linear map transposition is not enable and an uncanonical
          // `linear_function` instruction is encounter, emit a diagnostic.
          // FIXME(SR-11850): Finish support for linear map transposition.
          if (!EnableExperimentalLinearMapTransposition) {
            if (!lfi->hasTransposeFunction()) {
              astCtx.Diags.diagnose(
                  lfi->getLoc().getSourceLoc(),
                  diag::autodiff_conversion_to_linear_function_not_supported);
              errorOccurred = true;
            }
          }
          context.getLinearFunctionInstWorklist().push_back(lfi);
        }
```

This should fix `test/AutoDiff/SILOptimizer/differentiation_diagnostics.swift`.

`-Xllvm -enable-experimental-linear-map-transposition` must be added at the top of `test/AutoDiff/SILOptimizer/linear_function_canonicalization.sil`.